agent/proxycfg/state.go

```diff
@@ -965,6 +1025,241 @@ func (s *state) handleUpdateMeshGateway(u cache.UpdateEvent, snap *ConfigSnapsho
+		}
+		snap.Roots = roots
+	case u.CorrelationID == leafWatchID:
+		leaf, ok := u.Result.(*structs.IssuedCert)
+		if !ok {
+			return fmt.Errorf("invalid type for response: %T", u.Result)
+		}
+		snap.Leaf = leaf
+	case u.CorrelationID == serviceListWatchID:
+		// Store the list of services.
+		services, ok := u.Result.(*structs.IndexedServiceList)
+		if !ok {
+			return fmt.Errorf("invalid type for response: %T", u.Result)
+		}
+
+		serviceLists := make(map[string]map[structs.ServiceID]struct{})
+		for _, svc := range services.Services {
+			ns := svc.NamespaceOrDefault()
+			if _, ok := serviceLists[ns]; !ok {
+				serviceLists[ns] = make(map[structs.ServiceID]struct{})
+			}
+			sid := svc.ToServiceID()
+			serviceLists[ns][sid] = struct{}{}
+		}
+
+		snap.IngressGateway.ServiceLists = serviceLists
+
+		// Update our discovery chain watches.
+		if err := s.resetIngressUpstreamsAndDiscoveryWatches(snap); err != nil {
+			return err
+		}
+
+	case u.CorrelationID == ingressGatewayConfigWatchID:
+		configEntry, ok := u.Result.(*structs.ConfigEntryResponse)
+		if !ok {
+			return fmt.Errorf("invalid type for response: %T", u.Result)
+		}
+
+		ingressGateway, ok := configEntry.Entry.(*structs.IngressGatewayConfigEntry)
+		if !ok {
+			return fmt.Errorf("invalid type for config entry: %T", configEntry.Entry)
+		}
+
+		snap.IngressGateway.Config = ingressGateway
+
+		// Update our discovery chain watches.
+		if err := s.resetIngressUpstreamsAndDiscoveryWatches(snap); err != nil {
+			return err
+		}
+
+	case strings.HasPrefix(u.CorrelationID, "discovery-chain:"):
+		resp, ok := u.Result.(*structs.DiscoveryChainResponse)
+		if !ok {
+			return fmt.Errorf("invalid type for response: %T", u.Result)
+		}
+		svc := strings.TrimPrefix(u.CorrelationID, "discovery-chain:")
+		snap.IngressGateway.DiscoveryChain[svc] = resp.Chain
+
+		if err := s.resetIngressWatchesFromChain(svc, resp.Chain, snap); err != nil {
+			return err
+		}
+
+	case strings.HasPrefix(u.CorrelationID, "upstream-target:"):
+		resp, ok := u.Result.(*structs.IndexedCheckServiceNodes)
+		if !ok {
+			return fmt.Errorf("invalid type for response: %T", u.Result)
+		}
+		correlationID := strings.TrimPrefix(u.CorrelationID, "upstream-target:")
+		targetID, svc, ok := removeColonPrefix(correlationID)
+		if !ok {
+			return fmt.Errorf("invalid correlation id %q", u.CorrelationID)
+		}
+
+		_, ok = snap.IngressGateway.WatchedUpstreamEndpoints[svc]
+		if !ok {
+			m := make(map[string]structs.CheckServiceNodes)
+			snap.IngressGateway.WatchedUpstreamEndpoints[svc] = m
+		}
+		snap.IngressGateway.WatchedUpstreamEndpoints[svc][targetID] = resp.Nodes
+
+	default:
+		s.logger.Warn("received unknown update event", "event", u)
+	}
+
+	return nil
+}
+
+func (s *state) resetIngressUpstreamsAndDiscoveryWatches(snap *ConfigSnapshot) error {
+	// Exit early if we don't have both the gateway config and service list.
+	if snap.IngressGateway.Config == nil || snap.IngressGateway.ServiceLists == nil {
```

Comment:
Should the `ServiceLists` conditional wait on `== nil` or `len()==0`?

Reply:
Removed this entire function as it is now computed in the state store.